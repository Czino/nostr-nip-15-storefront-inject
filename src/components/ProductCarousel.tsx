import { useNostrHooks } from 'nostr-hooks'
import { useCustomNdk } from '../hooks/useCustomNdk'
import { useGetProductEvents } from '../hooks/useGetProductEvents'
import { Error } from './Error'
import { ProductItem } from './ProductItem'
const DEFAULTS = {
    LIMIT: 8,
}

type Props = {
    pubkey?: string
    productUrl?: string
    showPrice?: boolean
    limit?: number
    imageProxy?: string
    relays?: string
}
export const ProductCarousel = ({
    pubkey = '',
    productUrl,
    showPrice = true,
    limit = DEFAULTS.LIMIT,
    imageProxy,
    relays = '',
}: Props) => {
    const [customNdk] = useCustomNdk({ relays: relays ? relays.split(',') : undefined })
    useNostrHooks(customNdk)
    const { productInfos, eose } = useGetProductEvents({ pubkey, limit })
    const availableProducts = productInfos.filter((productInfo) => productInfo?.quantity >= 0)
    if (!pubkey) return <Error>Product Carousel: Missing attribute data-pubkey</Error>

    return (
        <div>
            <div className="flex overflow-x-auto relative gap-4 w-full snap-x">
                {availableProducts.map((productInfo) => (
                    <div
                        key={productInfo.id}
                        className={['snap-center flex-shrink-0 w-3/4 px-4', 'md:w-3/5', 'lg:px-1 lg:w-1/6'].join(' ')}
                    >
                        <ProductItem
                            product={productInfo}
                            productUrl={productUrl}
                            showPrice={showPrice}
                            width={340}
                            imageProxy={imageProxy}
                        />
                    </div>
                ))}
            </div>
            {!eose && <div className="text-center">Loading...</div>}
            {eose && availableProducts.length === 0 && <div className="text-center">No products found</div>}
        </div>
    )
}
