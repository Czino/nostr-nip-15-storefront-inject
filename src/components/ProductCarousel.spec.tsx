import { render } from '@testing-library/react'
import { productEvent } from '../../test/data/eventData'
import { mockUseSubscribe } from '../setupTests'
import { ProductCarousel } from './ProductCarousel'

describe('ProductCarousel', () => {
    let base: DocumentFragment
    it('renders error when no pubkey is passed', () => {
        mockUseSubscribe.mockReturnValue({ events: [], eose: false })
        const { asFragment } = render(<ProductCarousel />)
        expect(asFragment()).toMatchSnapshot()
    })
    it('renders product carousel correctly', () => {
        mockUseSubscribe.mockReturnValue({ events: [productEvent], eose: true })

        base = render(<ProductCarousel pubkey="pubkey" />).asFragment()
        expect(base).toMatchSnapshot()
    })
    it('renders product carousel correctly without price', () => {
        mockUseSubscribe.mockReturnValue({ events: [productEvent], eose: true })

        const { asFragment } = render(<ProductCarousel pubkey="pubkey" showPrice="false" />)
        expect(asFragment()).toMatchDiffSnapshot(base)
    })
})
