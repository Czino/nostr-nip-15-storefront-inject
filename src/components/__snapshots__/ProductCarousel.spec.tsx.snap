// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCarousel renders error when no pubkey is passed 1`] = `
<DocumentFragment>
  <div
    class="bg-red-700 text-white p-4 rounded text-center"
  >
    Product Carousel: Missing attribute data-pubkey
  </div>
</DocumentFragment>
`;

exports[`ProductCarousel renders product carousel correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex overflow-x-auto relative gap-4 w-full snap-x"
    >
      <div
        class="snap-center flex-shrink-0 w-3/4 px-4 md:w-3/5 lg:px-1 lg:w-1/6"
      >
        <a
          class="grid grid-cols-1 gap-1"
          href="/product/Cat%20Balaclava%20-%20Cactus/0f1a3984-7a72-4c6f-b1dc-bc170b6c4d8e"
        >
          <div
            class="w-full relative aspect-[2/3] overflow-hidden"
          >
            <div
              class="absolute inset-0 flex justify-center items-center -z-1"
            >
              <svg
                class="animate-spin h-7 w-7 text-light-1"
                fill="none"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  class="opacity-25"
                  cx="12"
                  cy="12"
                  r="10"
                  stroke="currentColor"
                  stroke-width="4"
                />
                <path
                  class="opacity-75"
                  d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <img
              alt=""
              class="w-full h-full object-cover"
              src="https://f004.backblazeb2.com/file/plebeian-market/P_listing_BHQQ_media_3041b54d1a0b9de92aa91999214b8848bc6951fe7afb5f9b3ead59d69d03aaf7.jpeg"
              width="340"
            />
            <img
              alt=""
              class="object-cover w-full h-full absolute top-0 opacity-0 hover:opacity-100 transition-opacity duration-500"
              src="https://f004.backblazeb2.com/file/plebeian-market/P_listing_BHQQ_media_3041b54d1a0b9de92aa91999214b8848bc6951fe7afb5f9b3ead59d69d03aaf8.jpeg"
              width="340"
            />
          </div>
          <div
            class="px-2"
          >
            <h3
              class="text-xl font-bold"
            >
              Cat Balaclava - Cactus
            </h3>
            <div>
              USD21
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProductCarousel renders product carousel correctly without price 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -56,10 +56,13 @@
              <h3
                class=\\"text-xl font-bold\\"
              >
                Cat Balaclava - Cactus
              </h3>
+             <div>
+               USD21
+             </div>
            </div>
          </a>
        </div>
      </div>
    </div>"
`;
